ivy_tests/test_ivy/test_frontends/test_torch/test_linalg.py

```diff
@@ -529,3 +529,76 @@ def test_torch_svdvals(
         on_device=on_device,
         A=x,
     )
+
+
+# tensorinv
+@st.composite
+def _get_inv_square_matrices(draw):
+    dim_size = draw(helpers.ints(min_value=1, max_value=10))
+
+    batch_shape = draw(st.sampled_from([2, 4, 6, 8, 10]))
+
+    generated_shape = (dim_size,) * batch_shape
+    generated_ind = int(np.floor(len(generated_shape) / 2))
+
+    handpicked_shape, handpicked_ind = draw(
+        st.sampled_from([[(24, 6, 4), 1], [(8, 3, 6, 4), 2], [(6, 7, 8, 16, 21), 3]])
+    )
+
+    shape, ind = draw(
+        st.sampled_from(
+            [(generated_shape, generated_ind), (handpicked_shape, handpicked_ind)]
+        )
+    )
+
+    input_dtype = draw(
+        helpers.get_dtypes("float", index=1, full=False).filter(
+            lambda x: x not in ["float16", "bfloat16"]
+        )
+    )
+    invertible = False
+    while not invertible:
+        input = draw(
+            helpers.array_values(
+                dtype=input_dtype[0],
+                shape=shape,
+                min_value=-100,
+                max_value=100,
+            )
+        )
+        try:
+            np.linalg.inv(input)
+            invertible = True
+        except np.linalg.LinAlgError:
+            pass
+
+    return input_dtype, input, ind
+
+
+@handle_frontend_test(
+    fn_tree="torch.linalg.tensorinv",
+    dtype_input_ind=_get_inv_square_matrices()
```

Comment:
You don't need all this code to produce `dtype_input_ind`. There is a simpler way to do this. 
Have a look at how `linalg_inv` does it:
```
    dtype_and_x=helpers.dtype_and_values(
        available_dtypes=helpers.get_dtypes("float"),
        min_value=-1e4,
        max_value=1e4,
        small_abs_safety_factor=3,
        shape=helpers.ints(min_value=2, max_value=10).map(lambda x: tuple([x, x])),
    ).filter(lambda x: np.linalg.cond(x[1]) < 1 / sys.float_info.epsilon),
```
In your case, you'll need to use `dtype_values_axis` instead of `dtype_and_values`. This will give you back an axis that will be your `ind`. 
The `.filter` part is basically a check that makes sure the array will be invertible. So you'll need it too.
And finally, `tensorinv` should work on square and non-square matrices, so you won't need the `.map` part at the `shape` argument. But you will need to `.filter` the shape so that the product of the first and last dimensions is equal.

Reply:
I tried this first but couldn't achieve to pass the tests, so I used helper implemented for jax tensorinv. I'll work on it.

Reply:
Related PR #7399
I found useful to check this PR and asked assignee of it about his opinion. He wasn't a hundred percent sure, but I got a confirmation that he said: ```
>i don't think there is a built-in helper function that allows you to generate such values, if there is such a function maybe you could ask your reviewer to point you towards it ?```

I tried `dtype_and_values_axis` helper, and It didn't work. I think it is not possible to control the shape of the tensor without drawing the axis and vice versa. If I filter the shape, the axis should be half of the length of the shape. If I filter arrays and their shape, again, can't control drew axis. 

```py
 np.prod(ivy.shape(x)[:int(len(ivy.shape(x))//2)]) == np.prod(ivy.shape(x)[int(len(ivy.shape(x))//2):]
 ```
As you may see, this has no relation with axis we need. I also checked arguments related to axis, none gives us to define it with respect to shape of drawn array.(`valid_axis` is just being sure that given ind, in the range of the shape dims.)

I also tried to solve it by using dtypes_values and set axis half of the shape inside test_linalg_tensorinv. Again, couldn't check product condition.

Sorry if it becomes a bit long.

Reply:
It can be done like this 
'''
    dtype_input_ind=helpers.dtype_values_axis(
        available_dtypes=helpers.get_dtypes("float"),
        min_value=-1e4,
        max_value=1e4,
        small_abs_safety_factor=3,
        shape=helpers.get_shape(min_num_dims=1).filter(
            lambda x:  np.prod(x[:int(len(x)//2)]) == np.prod(x[int(len(x)//2):])
        ),
        valid_axis=True,
        force_int_axis=True,
    ).filter(lambda x: np.linalg.cond(x[1]) < 1 / sys.float_info.epsilon),
'''

Reply:
Hi again, as I mentioned earlier, I already tried that. In this case, our drawn axis should be equal to `len(ivy.shape(x))//2`. However, the axis will be a random integer between `len(ivy.shape(x))>axis>1` but we need the mid-point of the shape drawn that fit the product condition. 
We have three conditions here, two depends on each other (In this case third condition not satisfied.):
1. Invertibility 
> [x]
2. Product condition 
>[x]
3. Axis should be the one that satisfy the product condition. 
> [ ]

Other option I tried is that using `dtype_and_values` helper with this setup and setting axis to the midpoint of the shape inside the test function. This time, our test handler can't check ind values.

Reply:
Yeah sorry I haven't gotten into depth with this function so I had misunderstood the 3rd condition. 
What do you mean "This time, our test handler can't check ind values."? If I understand correctly, the function itself forces the axis to be a certain integer given the shape of the input array, otherwise it produces an error. So it's okay to have the test use only this certain axis value. Alternatively, you could first draw a random integer for the axis and then draw the array filtering the shape based on that integer.

Reply:
You wouldn't do that inside `handle_frontend_test`. You could create an `st.composite` function where you first draw `ind` and then the `dtype_input` using the drawn value.

Reply:
Sorry if I am wrong, but there is an already defined `st.composite` function to exactly fit this function, as discussed at PR #7399.  What will be different from removing the jax helper I use right now from the same with different name? In case the problem is repetition, adding jax helper to the `helpers` and importing may help? The one used in there is following original test function implementation in the c++ code. I'm just asking what we're trying to avoid to understand and do better?

Reply:
The issue I see with the helper in the PR you have referenced is that it handpicks some shapes and indexes. It doesn't really randomly generate them through the hypothesis module. This means that the testing is limited and will not cover a wide range of possible inputs.

Reply:
I'm proposing something like:
```
@st.composite
def _tensorinv_helper(draw):
ind = draw(helpers.ints(min_value=1, max_value=10))
 dtype, input = draw(helpers.dtype_values_axis(
        available_dtypes = helpers.get_dtypes("float"),
        min_value=-1e4,
        max_value=1e4,
        small_abs_safety_factor=3,
        shape=helpers.get_shape(min_num_dims=1).filter(
            lambda x:  np.prod(ivy.shape(x)[:ind]) == np.prod(ivy.shape(x)[ind:])
        ).filter(lambda x: np.linalg.cond(x[1]) < 1 / sys.float_info.epsilon))
```

Reply:
Our preferred approach is to exhaustively test each function using `hypothesis`, which is indeed different than what Pytorch is doing.

Reply:
> I'm proposing something like:

I just used this one, here's the result:
![pycharm64_v9bD96fbgH](https://user-images.githubusercontent.com/51514616/210078558-2da5470f-26e6-4e52-972e-83516be545e3.png)
It draws arrays, but not tensorlike. 
And here is the original torch tensorinv test, written in python:
![clion64_M7ZJiOP77v](https://user-images.githubusercontent.com/51514616/210078789-f2f31f97-4b69-4030-8111-30c85dbc523b.png)

I got what I need to focus. Furthermore, I'll try to increase test coverage If I can do. But as far as I see, this function requires a very specific set of inputs to test. 
Also, I have a question about invertibility condition defined with ` np.linalg.cond(x[1]) < 1 / sys.float_info.epsilon`: Is this non-zero determinant imply invertibility for multidimensional tensors too?

Reply:
You need to give `input=x[0]` not `input=x` to the `test_frontend_function`. I think this will solve the TypeError.

Reply:
Yes it checks invertibility for any multidimensional array.

Reply:
> You need to give `input=x[0]` not `input=x` to the `test_frontend_function`. I think this will solve the TypeError.

`x[0]` means we don't pass a tensor to the function. We got that error because the function got a one dimensional array, so asked for a tensor. I solved `TypeError` by setting `min dim=2`. This is the error we need to solve.

```py
E   hypothesis.errors.Unsatisfiable: Unable to satisfy assumptions of test_torch_tensorinv.
```

I also removed invertibility filter and just tested with product condition. Why did I do that? Because I think the condition is too complicated to filter. And the error above is still present.

Please correct me if I am wrong. The current implementation of helper draws random arrays with the random number of dimensions we set, then we try to filter the shape that satisfies our product condition. So from that many random arrays, it is very unlikely to find a few that fit our product condition. Because it sets, both dimensions and the array sizes randomly. 

Why not filtering array (also dimensions) sizes  ? -Because those helpers get random values, and filtering them to find few whose shapes obey the rule is really hard. (Even if it works, probably it will create very few tensors (maybe none) that we need. A lot of randomness for very particular outcome.)

How can we generate such rules that produce tensors with the given rule? - That is the question we need to find an answer. 

To create such a rule, the length of the shape (number of dimensions) should be defined first. Because our helper function should know how many numbers required to decide factorization. 

Then we need to set the result of product to a certain number.  That depends on previous step.
(Alternatively, we can try to start with this step.  We need a number that have at least `x`  distinct factors so that we can set half of the shape length smaller than x. Even that require further filtering.)

Finally, we can fill each half of the shape with factorization of the number with the required number of factors. 

At last, we can fill the elements that will create invertible sub-tensors and tensors.

I am still not sure if this is the appropriate way to test this function. Most of the `linalg` functions have very simple input sets, unlike `tensorinv`. Again, I am kinda feel confident about none of those helpers around can help us with this process without creating a nested logic. All we can do is just using a combination of those helpers to have maybe just a bit bigger input pool than the `_get_inv_square_matrices` helper already has.

Reply:
> That is the question we need to find an answer.

I wrote a shape generation block that satisfy product rule and appropriate ind. However, filtering invertibility is not working as expected.

Reply:
This is an interesting problem. I haven't come across such a case in the past so thanks for sharing your thought process.
Your approach seems to be going in the right direction. What is the problem with invertibility?